cmd/collect_signals/worker.go

```diff
@@ -0,0 +1,155 @@
+	logger := w.logger.With(
+		zap.Int32("shard_id", req.GetShardNum()),
+		zap.Time("job_time", req.GetJobTime().AsTime()),
+		zap.String("filename", filename),
+	)
+	logger.Info("Processing shard")
+
+	// Prepare the output writer
+	extras := []string{}
+	if w.s != nil {
+		extras = append(extras, w.scoreColumnName)
+	}
+	var output bytes.Buffer
+	out := signalio.CsvWriter(&output, w.c.EmptySets(), extras...)
+
+	// Iterate through the repos in this shard.
+	for _, repo := range req.GetRepos() {
+		rawURL := repo.GetUrl()
+		if rawURL == "" {
+			logger.Warn("Skipping empty repo URL")
+			continue
+		}
+
+		// Create a logger for this repo.
+		repoLogger := logger.With(zap.String("repo", rawURL))
+		repoLogger.Info("Processing repo")
+
+		// Parse the URL to ensure it is a URL.
+		u, err := url.Parse(rawURL)
+		if err != nil {
+			// TODO: record a metric
+			repoLogger.With(zap.Error(err)).Warn("Failed to parse repo URL")
+			continue
+		}
+		ss, err := w.c.Collect(ctx, u)
+		if err != nil {
+			if errors.Is(err, collector.ErrUncollectableRepo) {
+				repoLogger.With(zap.Error(err)).Warn("Repo is uncollectable")
+				continue
+			}
+			return fmt.Errorf("failed during signal collection: %w", err)
+		}
+
+		// If scoring is enabled, prepare the extra data to be output.
+		extras := []signalio.Field{}
+		if w.s != nil {
+			f := signalio.Field{
+				Key:   w.scoreColumnName,
+				Value: fmt.Sprintf("%.5f", w.s.Score(ss)),
+			}
+			extras = append(extras, f)
+		}
+
+		// Write the signals to storage.
+		if err := out.WriteSignals(ss, extras...); err != nil {
+			return fmt.Errorf("failed writing signals: %w", err)
+		}
+	}
+
+	// Write to the canonical bucket last. The presence of the file indicates
+	// the job was completed. See scorecard's worker package for details.
+	if err := data.WriteToBlobStore(ctx, bucketURL, filename, output.Bytes()); err != nil {
+		return fmt.Errorf("error writing to blob store: %w", err)
+	}
+
+	logger.Info("Shard written successfully")
+
+	return nil
+}
+
+func getScorer(logger *zap.Logger, scoringEnabled bool, scoringConfigFile string) (*scorer.Scorer, error) {
+	if !scoringEnabled {
+		logger.Info("Scoring: disabled")
+		return nil, nil
+	}
+	if scoringConfigFile == "" {
+		logger.Info("Scoring: using default config")
+		return scorer.FromDefaultConfig(), nil
+	}
+	logger.With(zap.String("filename", scoringConfigFile)).Info("Scoring: using config file")
+
+	f, err := os.Open(scoringConfigFile)
+	if err != nil {
+		return nil, fmt.Errorf("opening config: %w", err)
+	}
+	defer f.Close()
+
+	s, err := scorer.FromConfig(scorer.NameFromFilepath(scoringConfigFile), f)
+	if err != nil {
+		return nil, fmt.Errorf("from config: %w", err)
+	}
+	return s, nil
+}
+
+func NewWorker(ctx context.Context, logger *zap.Logger, scoringEnabled bool, scoringConfigFile, scoringColumn string, collectOpts []collector.Option) (*collectWorker, error) {
+	logger.Info("Initializing worker")
+	logger.Debug("Creating collector")
```

Comment:
Is this debug log needed if we have an info log right before this?

Reply:
Fixed.